language/transaction-builder-generator/src/python3.rs

```diff
@@ -0,0 +1,139 @@
+use std::io::{Result, Write};
+
+pub fn output(out: &mut dyn Write, abis: &[ScriptABI]) -> Result<()> {
+    output_preambule(out)?;
+    for abi in abis {
+        output_builder(out, abi)?;
+    }
+    Ok(())
+}
+
+fn output_preambule(out: &mut dyn Write) -> Result<()> {
+    writeln!(
+        out,
+        r#"# pyre-ignore-all-errors
+import libra_types as libra
+import typing
+import serde_types as st"#
+    )
+}
+
+fn output_builder(out: &mut dyn Write, abi: &ScriptABI) -> Result<()> {
+    writeln!(
+        out,
+        "\ndef encode_{}_script({}) -> libra.Script:",
+        abi.name(),
+        [
+            quote_type_parameters(abi.ty_args()),
+            quote_parameters(abi.args()),
+        ]
+        .concat()
+        .join(", ")
+    )?;
+    writeln!(out, "{}", quote_doc(abi.doc()))?;
+    writeln!(
+        out,
+        r#"    return libra.Script(
+        code={},
+        ty_args=[{}],
+        args=[{}],
+    )"#,
+        quote_code(abi.code()),
+        quote_type_arguments(abi.ty_args()),
+        quote_arguments(abi.args()),
+    )?;
+    Ok(())
+}
+
+fn quote_doc(doc: &str) -> String {
+    let s: Vec<_> = doc.splitn(2, |c| c == '.').collect();
+    if s.len() <= 1 || s[1].is_empty() {
+        format!("    \"\"\"{}.\"\"\"", s[0])
+    } else {
+        format!(
+            r#"    """{}.
+
+{}    """"#,
+            s[0],
+            textwrap::indent(&textwrap::fill(&textwrap::dedent(s[1]), 86), "    "),
+        )
+    }
+}
+
+fn quote_type_parameters(ty_args: &[TypeArgumentABI]) -> Vec<String> {
+    ty_args
+        .iter()
+        .map(|ty_arg| format!("{}: libra.TypeTag", ty_arg.name()))
+        .collect()
+}
+
+fn quote_parameters(args: &[ArgumentABI]) -> Vec<String> {
+    args.iter()
+        .map(|arg| format!("{}: {}", arg.name(), quote_type(arg.type_tag())))
+        .collect()
+}
+
+fn quote_code(code: &[u8]) -> String {
+    format!(
+        "bytes([{}])",
+        code.iter()
+            .map(|x| format!("{}", x))
+            .collect::<Vec<_>>()
+            .join(", ")
+    )
+}
+
+fn quote_type_arguments(ty_args: &[TypeArgumentABI]) -> String {
+    ty_args
+        .iter()
+        .map(|ty_arg| ty_arg.name().to_string())
+        .collect::<Vec<_>>()
+        .join(", ")
+}
+
+fn quote_arguments(args: &[ArgumentABI]) -> String {
+    args.iter()
+        .map(|arg| make_transaction_argument(arg.type_tag(), arg.name()))
+        .collect::<Vec<_>>()
+        .join(", ")
+}
+
+fn quote_type(type_tag: &TypeTag) -> String {
+    use TypeTag::*;
+    match type_tag {
+        Bool => "st.bool".into(),
+        U8 => "st.uint8".into(),
+        U64 => "st.uint64".into(),
+        U128 => "st.uint128".into(),
+        Address => "libra.AccountAddress".into(),
+        Vector(type_tag) => match type_tag.as_ref() {
+            U8 => "bytes".into(),
+            _ => format!("typing.Sequence[{}]", quote_type(type_tag.as_ref())),
+        },
+
+        Struct(_) | Signer => panic!("not supported"),
```

Comment:
Might want this error message to imply that this is banned by Move--the current message sounds like this just might not be feature supported by the codegen.

Reply:
Same with other "not supported"'s on `Struct/Signer`

Reply:
Sure, I can do that.

For the records, I'm still curious why we allow complex types as type arguments but not complex values as regular arguments.

Reply:
The following invariant is important for Move's security model: A record type `R` declared in module `M` can only be created by a `Pack` instruction inside a procedure of `M`.  If record values can be passed into a transaction script, this invariant doesn't hold.

The use of types is not privileged in the same--any module can declare a procedure or type that uses any type declared in another module (as long it doesn't create a circular dependency).

Reply:
Said differently: producing a value of a particular type may be privileged, but producing a type of a particular kind never is.

Reply:
Thanks for the explanation! Based on what you're saying, nothing prevents us from having arbitrarily nested objects in TransactionArgument as long as they are benign enough (say vectors of vectors). So I understand we should probably plan for it in terms of code generation and LCS stack overflow.